mackerel-plugin-openldap/lib/openldap.go

```diff
@@ -0,0 +1,355 @@
+	Prefix             string
+	UseTLS             bool
+	InsecureSkipVerify bool
+	TargetHost         string
+	BindDn             string
+	ReplBase           string
+	ReplMasterHost     string
+	ReplMasterUseTLS   bool
+	ReplMasterBind     string
+	ReplMasterPass     string
+	ReplLocalBind      string
+	ReplLocalPass      string
+	BindPasswd         string
+	Tempfile           string
+	l                  *ldap.Conn
+}
+
+func transformKeyName(key string) string {
+	results := cnRegexp.FindStringSubmatch(key)
+	if len(results) < 2 {
+		return ""
+	}
+	return strings.Replace(strings.ToLower(strings.TrimSpace(results[1])), " ", "_", -1)
+}
+
+func getStats(sr *ldap.SearchResult, prefix string) map[string]float64 {
+	stat := make(map[string]float64)
+	for _, entry := range sr.Entries {
+		for _, attr := range entry.Attributes {
+			key := prefix + transformKeyName(entry.DN) + "_" + attr.Name
+			value, err := strconv.ParseFloat(entry.GetAttributeValue(attr.Name), 64)
+			if err != nil {
+				logger.Warningf("Failed to parse value. %s", err)
+			}
+			stat[key] = value
+		}
+	}
+	return stat
+}
+
+func fetchOpenldapMetrics(l *ldap.Conn, base, prefix string, attrs []string) (map[string]float64, error) {
+	searchRequest := ldap.NewSearchRequest(base, ldap.ScopeWholeSubtree, ldap.NeverDerefAliases, 0, 0, false, "(&(objectClass=*))", attrs, nil)
+	sr, err := l.Search(searchRequest)
+	if err != nil {
+		logger.Errorf("Failed to ldap search. %s.", err)
+		return nil, err
+	}
+	stat := getStats(sr, prefix)
+	return stat, nil
+}
+
+func mergeStat(dst, src map[string]float64) {
+	for k, v := range src {
+		dst[k] = v
+	}
+}
+
+func latestCSN(sr *ldap.SearchResult) (time.Time, error) {
+	var res time.Time
+	for _, entry := range sr.Entries {
+		csns := make([]time.Time, len(entry.Attributes))
+		for i, attr := range entry.Attributes {
+			v := entry.GetAttributeValue(attr.Name)
+			t, err := time.Parse("20060102150405.999999Z", v[0:strings.Index(v, "#")])
+			if err != nil {
+				return res, err
+			}
+			csns[i] = t
+		}
+		sort.Slice(csns, func(i, j int) bool {
+			return csns[i].UnixNano() > csns[j].UnixNano()
+		})
+		res = csns[0]
+		return res, nil
+	}
+	return res, errors.New("not found CSN")
+}
+
+func getLatestCSN(host, base, bind, passwd string, useTLS, insecureSkipVerify bool) (time.Time, error) {
+	var l *ldap.Conn
+	var err error
+	var res time.Time
+	if useTLS {
+		l, err = ldap.DialTLS("tcp", host, &tls.Config{InsecureSkipVerify: insecureSkipVerify})
+	} else {
+		l, err = ldap.Dial("tcp", host)
+	}
+	err = l.Bind(bind, passwd)
+	if err != nil {
+		logger.Errorf("Failed to Bind %s, err: %s", bind, err)
+		return res, err
+	}
+	searchRequest := ldap.NewSearchRequest(base, ldap.ScopeBaseObject, ldap.NeverDerefAliases, 0, 0, false, "(&(objectClass=*))", []string{"ContextCSN"}, nil)
+	sr, err := l.Search(searchRequest)
+	l.Close()
+	if err != nil {
+		logger.Errorf("Failed to ldap search. %s.", err)
+		return res, err
+	}
+	return latestCSN(sr)
```

Comment:
Since multiple CSNs are obtained like https://github.com/openldap/openldap/blob/master/tests/scripts/test058-syncrepl-asymmetric#L1211-L1216 ,
sort them and get the newest value.
And the difference between the master and slave CSN values is used as a metric.